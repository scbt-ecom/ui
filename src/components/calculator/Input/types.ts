import type { InputProps as MuiInputProps } from "@mui/material";

type InputBaseProps = Pick<MuiInputProps, "className">;

export interface InputProps extends InputBaseProps, MuiInputProps {
  defaultValue: number;
  min: number;
  max: number;
}
