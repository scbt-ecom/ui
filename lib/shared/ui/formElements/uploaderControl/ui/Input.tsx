import { forwardRef, type Ref } from 'react'
import { type DropzoneRootProps } from 'react-dropzone'
import { type FieldError, type FieldValues, type Path } from 'react-hook-form'
import { type TUploaderClasses } from '../model/types'
import { Icon } from '$/shared/ui/icon'
import { cn } from '$/shared/utils'

export interface IFileInputProps<T extends FieldValues> {
  dropzoneState: DropzoneRootProps
  disabled?: boolean
  error?: FieldError
  classes?: TUploaderClasses
  name?: Path<T>
}

export const Input = forwardRef(function Input<T extends FieldValues>(
  { dropzoneState, classes, disabled, error, name }: IFileInputProps<T>,
  ref: Ref<HTMLDivElement>
) {
  const dropzoneProps = dropzoneState.getRootProps()
  return (
    <div
      className={cn(
        'flex h-[64px] w-[476px] rounded-sm bg-[length:100%_100%] bg-no-repeat hover:border hover:border-primary-hover hover:bg-color-primary-tr-hover focus-visible:border focus-visible:border-primary-hover focus-visible:bg-color-primary-tr-hover focus-visible:outline-none active:border active:border-primary-hover active:bg-color-primary-tr-pressed [&:not(:hover)(:focus-visible)(:active)(:disabled)]:bg-[url("/static/files/border.svg")]',
        classes?.borderContent,
        { 'bg-color-primary-light-default opacity-50': disabled },
        { '[&:not(:hover)(:focus-visible)(:active)(:disabled)]:bg-[url("/static/files/borderError.svg")]': error }
      )}
      ref={ref}
      {...dropzoneProps}
    >
      <div className={cn('flex h-full w-full items-center justify-center gap-2', classes?.wrapperTextContent)}>
        <Icon name='files/upload' className={cn('text-icon-primary-default', classes?.uploaderIcon)} />
        <p className={cn('desk-body-regular-l select-none text-icon-primary-default', classes?.wrapperTextContent)}>
          Выберите файл
          <span className={cn('text-icon-blue-grey-600', classes?.selectFileTextSpan)}> или перетащите сюда</span>
        </p>
      </div>
      <input
        name={name}
        ref={dropzoneState.inputRef}
        disabled={disabled}
        {...dropzoneState.getInputProps()}
        className={cn(classes?.uploaderInput)}
      />
    </div>
  )
})
