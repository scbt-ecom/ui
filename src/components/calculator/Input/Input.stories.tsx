import type { Meta, StoryObj } from "@storybook/react";
import { Input } from ".";

const meta: Meta<typeof Input> = {
  title: "Calculator/Input",
  component: Input,
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Standard: Story = {
  args: {
    defaultValue: 5,
    max: 10,
    min: 0,
  },
};
