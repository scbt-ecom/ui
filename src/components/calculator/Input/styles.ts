import { styled } from "@mui/material/styles";

import TextField from "@mui/material/TextField";

export const CustomTextField = styled(TextField)({
  "& .MuiOutlinedInput-notchedOutline": {
    border: "none",
  },
  input: {
    "&[type=number]": {
      MozAppearance: "textfield",
    },
    "&::-webkit-outer-spin-button": {
      WebkitAppearance: "none",
      margin: 0,
    },
    "&::-webkit-inner-spin-button": {
      WwebkitAppearance: "none",
      margin: 0,
    },
  },
});
