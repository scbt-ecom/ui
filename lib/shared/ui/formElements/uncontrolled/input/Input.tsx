import { forwardRef, useId } from 'react'
import { FieldAttachment } from '../../ui'
import { type DeepPartial } from '$/shared/types'
import { cn } from '$/shared/utils'

export type InputBaseClasses = {
  container?: string
  input?: string
  label?: string
}

type FieldAttachmentProps = React.ComponentPropsWithoutRef<typeof FieldAttachment>

type ExternalHandlers = {
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void
  onFocus?: (event: React.FocusEvent<HTMLDivElement>) => void
  onBlur?: (event: React.FocusEvent<HTMLDivElement>) => void
}

export type InputBaseProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, 'placeholder' | 'size'> & {
  /**
   * Дополнительные стили внутренних компонентов
   */
  classes?: InputBaseClasses
  /**
   * Отображаемый placeholder
   */
  label: string
  /**
   * Отображение не валидного поля
   */
  invalid?: boolean
  /**
   * Свойства дополнительной иконки
   */
  attachmentProps?: DeepPartial<FieldAttachmentProps>
  /**
   * Рендер дополнительных значений вместо базового ввода
   */
  renderValues?: () => React.JSX.Element | null
  /**
   * Дополнительные хендлеры
   */
  externalHandlers?: ExternalHandlers
}

export const InputBase = forwardRef<HTMLInputElement, InputBaseProps>(
  ({ label, value, invalid, disabled, classes, renderValues, attachmentProps, externalHandlers, onChange, ...props }, ref) => {
    const id = useId()

    const { onChange: externalOnChange, ...restHandlers } = externalHandlers || {}

    const onValueChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      if (onChange) onChange(event)
      if (externalOnChange) externalOnChange(event)
    }

    return (
      <div
        ref={renderValues ? ref : undefined}
        className={cn(
          'relative flex w-full gap-x-4 rounded-sm bg-color-blue-grey-100',
          '[&_label]:focus-within:top-[9px] [&_label]:focus-within:translate-y-0',
          '[&_label]:focus-within:desk-body-regular-s',
          '[&_label]:focus-within:text-color-tetriary',
          'hover:bg-color-blue-grey-200',
          'border-[1px] border-transparent focus-within:border-blue-grey-800',
          {
            'border-secondary-default': invalid
          },
          classes?.container
        )}
      >
        {renderValues ? (
          <div
            onClick={props.onClick}
            onBlur={props.onBlur}
            className={cn(
              'peer desk-body-regular-l min-h-[54px] w-full bg-color-transparent [&_input]:px-4 [&_input]:pb-[6px] [&_input]:pt-[28px]',
              'text-color-dark outline-none placeholder:text-color-transparent',
              'flex flex-row flex-wrap gap-2'
            )}
          >
            {renderValues()}
          </div>
        ) : (
          <input
            {...props}
            {...restHandlers}
            onChange={onValueChange}
            disabled={disabled}
            value={value ?? ''}
            placeholder={label}
            aria-placeholder={label}
            ref={ref}
            id={id}
            className={cn(
              'peer desk-body-regular-l w-full bg-color-transparent px-4 pb-[9px] pt-[25px]',
              'text-color-dark outline-none placeholder:text-color-transparent',
              {
                hidden: Boolean(renderValues)
              },
              classes?.input
            )}
          />
        )}
        <label
          htmlFor={id}
          className={cn(
            'desk-body-regular-l pointer-events-none absolute left-4',
            'top-1/2 -translate-y-1/2 text-color-blue-grey-600 duration-100',
            {
              'desk-body-regular-s top-[9px] translate-y-0 text-color-tetriary': Boolean(value),
              'text-color-disabled': disabled
            },
            classes?.label
          )}
        >
          {label}
        </label>
        {attachmentProps && <FieldAttachment {...attachmentProps} />}
      </div>
    )
  }
)
InputBase.displayName = 'InputBase'
