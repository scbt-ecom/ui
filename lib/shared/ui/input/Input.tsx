import { forwardRef, useId } from 'react'
import { FieldAttachment } from '../formElements/ui'
import { type DeepPartial } from '$/shared/types'
import { cn } from '$/shared/utils'

type InputBaseClasses = {
  container?: string
  input?: string
  label?: string
}

type FieldAttachmentProps = React.ComponentPropsWithoutRef<typeof FieldAttachment>

export type InputBaseProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, 'placeholder'> & {
  /**
   * Дополнительные стили внутренних компонентов
   */
  classes?: InputBaseClasses
  /**
   * Отображаемый placeholder
   */
  label: string
  /**
   * Отображение не валидного поля
   */
  invalid?: boolean
  /**
   * Свойства дополнительной иконки
   */
  attachmentProps?: DeepPartial<FieldAttachmentProps>
  /**
   * Рендер дополнительных значений вместо базового ввода
   */
  renderValues?: () => React.JSX.Element | null
}

export const InputBase = forwardRef<HTMLInputElement, InputBaseProps>(
  ({ label, value, invalid, disabled, classes, renderValues, attachmentProps, ...props }, ref) => {
    const id = useId()

    return (
      <div
        ref={renderValues ? ref : undefined}
        className={cn(
          'relative flex w-full gap-x-4 rounded-sm bg-color-blue-grey-100',
          '[&_label]:focus-within:top-[9px] [&_label]:focus-within:translate-y-0',
          '[&_label]:focus-within:desk-body-regular-s',
          '[&_label]:focus-within:text-color-tetriary',
          'hover:bg-color-blue-grey-200',
          'border-[1px] border-transparent focus-within:border-blue-grey-800',
          {
            'border-secondary-default': invalid
          },
          classes?.container
        )}
      >
        {renderValues ? (
          <div
            onClick={props.onClick}
            onBlur={props.onBlur}
            className={cn(
              'peer desk-body-regular-l min-h-[54px] w-full bg-color-transparent px-4 pb-[6px] pt-[28px]',
              'text-color-dark outline-none placeholder:text-color-transparent',
              'flex flex-row flex-wrap gap-2'
            )}
          >
            {renderValues()}
          </div>
        ) : (
          <input
            {...props}
            disabled={disabled}
            value={value}
            placeholder={label}
            aria-placeholder={label}
            ref={ref}
            id={id}
            className={cn(
              'peer desk-body-regular-l w-full bg-color-transparent px-4 pb-[9px] pt-[25px]',
              'text-color-dark outline-none placeholder:text-color-transparent',
              classes?.input
            )}
          />
        )}
        <label
          htmlFor={id}
          className={cn(
            'desk-body-regular-l pointer-events-none absolute left-4',
            'top-1/2 -translate-y-1/2 text-color-blue-grey-600 duration-100',
            {
              'desk-body-regular-s top-[9px] translate-y-0 text-color-tetriary': Boolean(value),
              'text-color-disabled': disabled
            },
            classes?.label
          )}
        >
          {label}
        </label>
        {attachmentProps && <FieldAttachment {...attachmentProps} />}
      </div>
    )
  }
)
InputBase.displayName = 'InputBase'
