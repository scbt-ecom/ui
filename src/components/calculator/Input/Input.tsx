import type { ChangeEvent } from "react";
import { useState } from "react";

import type { InputProps } from "./types";

import { CustomTextField } from "./styles";

export const Input = ({ defaultValue, min, max }: InputProps) => {
  const [value, setValue] = useState(defaultValue);

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.target.value === "" ? 0 : Number(event.target.value));
  };

  const handleBlur = () => {
    if (value < min) {
      setValue(min);
    } else if (value > max) {
      setValue(max);
    }
  };

  return (
    <CustomTextField
      value={value}
      size="small"
      onChange={handleInputChange}
      id="outlined-start-adornment"
      onBlur={handleBlur}
      inputProps={{
        step: null,
        min,
        max,
        type: "number",
        "data-testid": "TextFieldTestId",
      }}
    />
  );
};
