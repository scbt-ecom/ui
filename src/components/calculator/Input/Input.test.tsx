import { render, screen, fireEvent } from "@testing-library/react";
import { Input } from "./Input";

describe("input Component", () => {
  it("renders input", () => {
    expect.hasAssertions();
    render(<Input defaultValue={5} min={0} max={10} />);

    const input = screen.getByTestId("TextFieldTestId");
    expect(input).toBeInTheDocument();
    expect(input).toMatchSnapshot();
  });

  it("entering a value within the allowed range", () => {
    expect.hasAssertions();
    render(<Input defaultValue={5} min={0} max={10} />);
    const inputField = screen.getByTestId("TextFieldTestId") as HTMLInputElement;

    fireEvent.change(inputField, { target: { value: "5" } });
    expect(inputField.value).toBe("5");
  });

  it("entering a value greater than within the permitted range", () => {
    expect.hasAssertions();
    render(<Input defaultValue={5} min={0} max={10} />);
    const inputField = screen.getByTestId("TextFieldTestId") as HTMLInputElement;

    fireEvent.change(inputField, { target: { value: "45" } });
    fireEvent.blur(inputField);
    expect(inputField.value).toBe("10");
  });

  it("entering a value less than within the allowed range", () => {
    expect.hasAssertions();
    render(<Input defaultValue={5} min={0} max={10} />);
    const inputField = screen.getByTestId("TextFieldTestId") as HTMLInputElement;

    fireEvent.change(inputField, { target: { value: "-45" } });
    fireEvent.blur(inputField);
    expect(inputField.value).toBe("0");
  });

  it("entering non-digits", () => {
    expect.hasAssertions();
    render(<Input defaultValue={5} min={0} max={10} />);
    const inputField = screen.getByTestId("TextFieldTestId") as HTMLInputElement;

    fireEvent.change(inputField, { target: { value: "Value" } });
    fireEvent.blur(inputField);
    expect(inputField.value).toBe("0");
  });
});
