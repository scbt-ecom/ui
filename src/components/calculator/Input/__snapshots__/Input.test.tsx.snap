// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input Component renders input 1`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
  data-testid="TextFieldTestId"
  id="outlined-start-adornment"
  max="10"
  min="0"
  type="number"
  value="5"
/>
`;
